File: sdk/ts/src/utils/logger.ts
export type LogHandler = (level: string, message: string, ...args: any[]) => void;

export enum LogLevel {
    TRACE = 0,
    DEBUG = 1,
    INFO = 2,
    WARN = 3,
    ERROR = 4,
    NONE = 5
}

export class Logger {
    private readonly level: LogLevel;
    private readonly handler: LogHandler;

    constructor(options?: { handler?: LogHandler, level?: string }) {
        const envLevel = options?.level || process.env.NEXO_LOG?.toUpperCase() || 'ERROR';
        this.level = this.parseLevel(envLevel);
        this.handler = options?.handler || this.defaultHandler;
    }

    private parseLevel(lvl: string): LogLevel {
        switch (lvl) {
            case 'TRACE': return LogLevel.TRACE;
            case 'DEBUG': return LogLevel.DEBUG;
            case 'INFO': return LogLevel.INFO;
            case 'WARN': return LogLevel.WARN;
            case 'ERROR': return LogLevel.ERROR;
            case 'OFF': return LogLevel.NONE;
            default: return LogLevel.ERROR;
        }
    }

    private defaultHandler(level: string, msg: string, ...args: any[]) {
        const timestamp = new Date().toISOString();
        const prefix = `[SDK] [${timestamp}] ${level}`;

        switch (level) {
            case 'ERROR': console.error(prefix, msg, ...args); break;
            case 'WARN': console.warn(prefix, msg, ...args); break;
            default: console.log(prefix, msg, ...args); break;
        }
    }

    trace(msg: string, ...args: any[]) {
        if (this.level <= LogLevel.TRACE) {
            this.handler('TRACE', msg, ...args);
        }
    }

    debug(msg: string, ...args: any[]) {
        if (this.level <= LogLevel.DEBUG) {
            this.handler('DEBUG', msg, ...args);
        }
    }

    info(msg: string, ...args: any[]) {
        if (this.level <= LogLevel.INFO) {
            this.handler('INFO', msg, ...args);
        }
    }

    warn(msg: string, ...args: any[]) {
        if (this.level <= LogLevel.WARN) {
            this.handler('WARN', msg, ...args);
        }
    }

    error(msg: string, ...args: any[]) {
        if (this.level <= LogLevel.ERROR) {
            this.handler('ERROR', msg, ...args);
        }
    }
}

File: sdk/py/src/nexo/utils/logger.py
from __future__ import annotations

import os
import sys
from datetime import datetime, timezone
from typing import Any, Callable, Optional

from ..config import DEFAULT_CONFIG

LogHandler = Callable[[str, str, Any], None]


class LogLevel:
    TRACE = 0
    DEBUG = 1
    INFO = 2
    WARN = 3
    ERROR = 4
    NONE = 5

    _names = {
        "TRACE": TRACE,
        "DEBUG": DEBUG,
        "INFO": INFO,
        "WARN": WARN,
        "ERROR": ERROR,
        "OFF": NONE,
    }

    @classmethod
    def parse(cls, level: str) -> int:
        return cls._names.get(level.upper(), cls.ERROR)


class Logger:
    def __init__(
        self,
        handler: Optional[LogHandler] = None,
        level: Optional[str] = None,
    ) -> None:
        env_level = level or os.environ.get("NEXO_LOG", "").upper() or DEFAULT_CONFIG.logger.level
        self._level = LogLevel.parse(env_level)
        self._handler: LogHandler = handler or self._default_handler

    def _default_handler(self, level: str, msg: str, *args: Any) -> None:
        timestamp = datetime.now(timezone.utc).isoformat()
        prefix = f"[SDK] [{timestamp}] {level}"
        if level == "ERROR":
            print(prefix, msg, *args, file=sys.stderr)
        elif level == "WARN":
            print(prefix, msg, *args, file=sys.stderr)
        else:
            print(prefix, msg, *args)

    def trace(self, msg: str, *args: Any) -> None:
        if self._level <= LogLevel.TRACE:
            self._handler("TRACE", msg, *args)

    def debug(self, msg: str, *args: Any) -> None:
        if self._level <= LogLevel.DEBUG:
            self._handler("DEBUG", msg, *args)

    def info(self, msg: str, *args: Any) -> None:
        if self._level <= LogLevel.INFO:
            self._handler("INFO", msg, *args)

    def warn(self, msg: str, *args: Any) -> None:
        if self._level <= LogLevel.WARN:
            self._handler("WARN", msg, *args)

    def error(self, msg: str, *args: Any) -> None:
        if self._level <= LogLevel.ERROR:
            self._handler("ERROR", msg, *args)
